File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.java
// [-7,-3,2,3,11]
// [49, 9, 4, 9, 121]
//.  l        r
//. [n, n, n, n, 121]

class Solution {
    public int[] sortedSquares(int[] nums) {
        int[] result = new int[nums.length];
        int left = 0;
        int right = nums.length - 1;
        int i = nums.length - 1;
        while(left <= right) {
            int leftSq = nums[left] * nums[left];
            int rightSq = nums[right] * nums[right];
            
            if(leftSq >= rightSq) {
                result[i] = leftSq;
                left += 1;
            } else {
                result[i] = rightSq;
                right -= 1;
            }
            
            i--;
        }
        return result;
    }
}

File: 977-squares-of-a-sorted-array/977-squares-of-a-sorted-array.py
'''
 16 1 0 9 100
 n      p
'''
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        sorted_sq = []
        i = 0 
        while i < len(nums):
            if nums[i] >= 0:
                break
            i += 1
        
        neg = i - 1
        pos = i
        
        while pos < len(nums) and neg >= 0:
            print(nums[pos], nums[neg])
            sq_pos = nums[pos] * nums[pos]
            sq_neg = nums[neg] * nums[neg]
            
            if sq_pos > sq_neg:
                sorted_sq.append(sq_neg)
                neg -= 1
            else:
                sorted_sq.append(sq_pos)
                pos += 1
        
        while pos < len(nums):
            sorted_sq.append(nums[pos] * nums[pos])
            pos += 1
        while neg >= 0:
            sorted_sq.append(nums[neg] * nums[neg])
            neg -= 1
            
        return sorted_sq